File: scripts/verify_widget.py
#!/usr/bin/env python3
"""verify_widget.py — браузерная проверка живых виджетов Сигмы.
Грузит страницу, ждёт монтирования, ловит console-ошибки, двигает каждый
ползунок (min→max), проверяет факт перерисовки canvas, скриншотит до/после.
Верификация + инструмент design-review.

Запуск:  /root/.venv/bin/python scripts/verify_widget.py <url> <outDir>
"""
import json
import os
import sys

from playwright.sync_api import sync_playwright

url = sys.argv[1] if len(sys.argv) > 1 else "https://sigma.fmin.xyz/story_pca.html"
out_dir = sys.argv[2] if len(sys.argv) > 2 else "/tmp/sigverify"
os.makedirs(out_dir, exist_ok=True)
slug = url.rstrip("/").split("/")[-1].replace(".html", "") or "page"

errors = []
report = {"url": url, "widgets": 0, "errors": [], "results": []}

CANVAS_HASH = """el => { const c = el.querySelector('canvas'); if(!c) return null;
  try { const u = c.toDataURL(); return u.length + ':' + u.slice(-80); } catch(e){ return 'tainted'; } }"""

with sync_playwright() as p:
    browser = p.chromium.launch()
    page = browser.new_page(viewport={"width": 1100, "height": 1400}, device_scale_factor=2)
    page.on("console", lambda m: errors.append(m.text) if m.type == "error" else None)
    page.on("pageerror", lambda e: errors.append("PAGEERROR: " + e.message))
    page.goto(url, wait_until="networkidle", timeout=60000)
    page.wait_for_timeout(800)

    widgets = page.locator(".sigma-int")
    n = widgets.count()
    report["widgets"] = n
    if n == 0:
        report["errors"].append("НЕТ смонтированных .sigma-int виджетов")

    for i in range(n):
        w = widgets.nth(i)
        w.scroll_into_view_if_needed()
        page.wait_for_timeout(500)
        r = {"index": i, "hasCanvas": False, "hasSvg": False, "reactsToSlider": None, "err": None}
        if w.locator(".sigma-int-err").count():
            r["err"] = w.locator(".sigma-int-err").first.text_content()
        r["hasCanvas"] = w.locator("canvas").count() > 0
        r["hasSvg"] = w.locator("svg.sigma-int-svg").count() > 0
        try:
            w.screenshot(path=f"{out_dir}/{slug}_w{i}_before.png")
        except Exception:
            pass

        ranges = w.locator("input[type=range]")
        nr = ranges.count()
        if nr > 0 and r["hasCanvas"]:
            handle = w.element_handle()
            h0 = handle.evaluate(CANVAS_HASH)
            rng = ranges.first
            mx = rng.get_attribute("max")
            mn = rng.get_attribute("min")
            for val in (mx, mn):
                rng.fill(str(val))
                rng.dispatch_event("input")
                page.wait_for_timeout(300)
            h1 = handle.evaluate(CANVAS_HASH)
            r["reactsToSlider"] = (h0 != h1)
            try:
                mid = round((float(mn) + float(mx)) / 2)
                rng.fill(str(mid))
                rng.dispatch_event("input")
                page.wait_for_timeout(300)
            except Exception:
                pass
            try:
                w.screenshot(path=f"{out_dir}/{slug}_w{i}_after.png")
            except Exception:
                pass
        report["results"].append(r)

    report["errors"].extend(errors)
    browser.close()

with open(f"{out_dir}/{slug}_report.json", "w") as f:
    json.dump(report, f, ensure_ascii=False, indent=2)
print(json.dumps(report, ensure_ascii=False, indent=2))

File: scripts/verify_widget.mjs
// verify_widget.mjs — браузерная проверка живых виджетов Сигмы.
// Грузит страницу, ждёт монтирования, ловит console-ошибки, двигает каждый
// ползунок (min→max), проверяет что виджет ПЕРЕРИСОВЫВАЕТСЯ (canvas меняется),
// делает скриншоты до/после. Это и верификация, и инструмент design-review.
//
// Запуск:  node scripts/verify_widget.mjs <url> <outDir> [widgetIndex]
//   node scripts/verify_widget.mjs https://sigma.fmin.xyz/story_pca.html /tmp/sigverify
import { chromium } from "playwright";
import fs from "fs";

const url = process.argv[2] || "https://sigma.fmin.xyz/story_pca.html";
const outDir = process.argv[3] || "/tmp/sigverify";
fs.mkdirSync(outDir, { recursive: true });
const slug = url.split("/").pop().replace(/\.html$/, "") || "page";

const browser = await chromium.launch();
const page = await browser.newPage({ viewport: { width: 1100, height: 1400 }, deviceScaleFactor: 2 });

const errors = [];
page.on("console", (m) => { if (m.type() === "error") errors.push(m.text()); });
page.on("pageerror", (e) => errors.push("PAGEERROR: " + e.message));

await page.goto(url, { waitUntil: "networkidle", timeout: 60000 });

// дождаться появления холстов/SVG внутри виджетов
const widgets = page.locator(".sigma-int");
const n = await widgets.count();
const report = { url, widgets: n, errors: [], results: [] };

if (n === 0) {
  report.errors.push("НЕТ смонтированных .sigma-int виджетов на странице");
}

// hash содержимого canvas (для проверки факта перерисовки)
async function canvasHash(handle) {
  return await handle.evaluate((el) => {
    const c = el.querySelector("canvas");
    if (!c) return null;
    try { return c.toDataURL().length + ":" + c.toDataURL().slice(-64); } catch (e) { return "tainted"; }
  });
}

for (let i = 0; i < n; i++) {
  const w = widgets.nth(i);
  await w.scrollIntoViewIfNeeded();
  await page.waitForTimeout(500);
  const r = { index: i, hasCanvas: false, hasSvg: false, reactsToSlider: null, err: null };
  const hErr = await w.locator(".sigma-int-err").count();
  if (hErr) { r.err = await w.locator(".sigma-int-err").first().textContent(); }
  r.hasCanvas = (await w.locator("canvas").count()) > 0;
  r.hasSvg = (await w.locator("svg.sigma-int-svg").count()) > 0;

  await w.screenshot({ path: `${outDir}/${slug}_w${i}_before.png` }).catch(() => {});

  // подвигать первый ползунок min→mid→max и проверить перерисовку
  const ranges = w.locator("input[type=range]");
  const nr = await ranges.count();
  if (nr > 0 && r.hasCanvas) {
    const handle = await w.elementHandle();
    const h0 = await canvasHash(handle);
    const rng = ranges.first();
    const max = await rng.getAttribute("max");
    const min = await rng.getAttribute("min");
    await rng.fill(String(max));
    await rng.dispatchEvent("input");
    await page.waitForTimeout(300);
    const h1 = await canvasHash(handle);
    await rng.fill(String(min));
    await rng.dispatchEvent("input");
    await page.waitForTimeout(300);
    const h2 = await canvasHash(handle);
    r.reactsToSlider = (h0 !== h1) || (h1 !== h2);
    await rng.fill(String(Math.round((+min + +max) / 2)));
    await rng.dispatchEvent("input");
    await page.waitForTimeout(300);
    await w.screenshot({ path: `${outDir}/${slug}_w${i}_after.png` }).catch(() => {});
  }
  report.results.push(r);
}

report.errors.push(...errors);
fs.writeFileSync(`${outDir}/${slug}_report.json`, JSON.stringify(report, null, 2));
console.log(JSON.stringify(report, null, 2));
await browser.close();
